python/tritonserver/_c/tritonserver_pybind.cc

```diff
@@ -1175,20 +1171,77 @@ class PyInferenceRequest
     ThrowIfError(TRITONSERVER_InferenceRequestSetDoubleParameter(
         triton_object_, key.c_str(), value));
   }
+
   void Cancel()
   {
     ThrowIfError(TRITONSERVER_InferenceRequestCancel(triton_object_));
   }
 
+  // Response management
+  void GetNextResponse(const py::object& py_future)
+  {
+    std::lock_guard lock(response_mu_);
 
- public:
-  std::unique_ptr<CallbackResource> request_callback_resource_{nullptr};
+    if (responses_.empty()) {
+      if (response_future_.get() != nullptr) {
+        throw AlreadyExistsError("cannot call GetNextResponse concurrently");
+      }
+      response_future_.reset(new py::object(py_future));
+    } else {
+      std::pair<std::shared_ptr<PyInferenceResponse>, const uint32_t>&
+          py_response = responses_.front();
+      PyFutureSetResult(py_future, py_response);
+      responses_.pop_front();
+    }
+  }
 
  private:
-  std::unique_ptr<PyResponseAllocator::CallbackResource>
-      allocator_callback_resource_{nullptr};
-  std::unique_ptr<PyInferenceResponse::CallbackResource>
-      response_callback_resource_{nullptr};
+  void AddNextResponse(
+      struct TRITONSERVER_InferenceResponse* response, const uint32_t flags)
+  {
+    std::shared_ptr<PyInferenceResponse> managed_ptr(nullptr);
+    if (response != nullptr) {
+      managed_ptr.reset(new PyInferenceResponse(response, true /* owned */));
+    }
+    std::pair<std::shared_ptr<PyInferenceResponse>, const uint32_t> py_response(
+        std::move(managed_ptr), std::move(flags));
+    {
+      std::lock_guard lock(response_mu_);
+      if (response_future_.get() == nullptr) {
+        responses_.push_back(std::move(py_response));
+        return;
+      }
+    }
+    {
+      py::gil_scoped_acquire gil;
+      std::unique_ptr<py::object> response_future_local(nullptr);
+      response_future_.swap(response_future_local);
+      PyFutureSetResult(*response_future_local, py_response);
+    }
+  }
+
+  void PyFutureSetResult(
+      const py::object& py_future,
+      std::pair<std::shared_ptr<PyInferenceResponse>, const uint32_t>&
+          py_response)
+  {
+    py::tuple py_res = py::make_tuple(
+        (py_response.first.get() ? py::cast(py_response.first) : py::none()),
+        py_response.second);
+    if (py::hasattr(py_future, "get_loop")) {
+      py_future.attr("get_loop")().attr("call_soon_threadsafe")(
+          py_future.attr("set_result"), std::move(py_res));
```

Comment:
neat! So this support async and non async?

Reply:
yes